FueledUtils/Core/Atomic.swift

```diff
@@ -0,0 +1,91 @@
+// Copyright © 2020, Fueled Digital Media, LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+
+infix operator .=: AssignmentPrecedence
+
+public final class AtomicValue<Value> {
+	private(set) var value: Value
+	private let lock = Lock()
+
+	public init(_ value: Value) {
+		self.value = value
+	}
+
+	public static func .= (atomicValue: AtomicValue, value: Value) {
+		atomicValue.modify { $0 = value }
+	}
+
+	public func modify(_ modify: (inout Value) -> Void) {
+		self.lock.lock()
+		defer {
+			self.lock.unlock()
+		}
+		modify(&self.value)
+	}
+
+	public func withValue(_ getter: (Value) -> Void) {
+		self.lock.lock()
+		defer {
+			self.lock.unlock()
+		}
+		getter(self.value)
+	}
+}
+
+@propertyWrapper
+public struct Atomic<Value> {
+	private let atomicValue: AtomicValue<Value>
+
+	public init(_ value: Value) {
+		self.atomicValue = AtomicValue(value)
+	}
+
+	public init(wrappedValue value: Value) {
+		self.init(value)
+	}
+
+	public var wrappedValue: Value {
+		self.atomicValue.value
+	}
+
+	public mutating func modify(_ modify: (inout Value) -> Void) {
+		self.atomicValue.modify(modify)
+	}
+
+	public mutating func withValue(_ getter: (Value) -> Void) {
+		self.atomicValue.withValue(getter)
+	}
+
+	public static func .= (atomicValue: inout Atomic, value: Value) {
+		atomicValue.atomicValue .= value
+	}
+}
+
+@propertyWrapper
+public struct LazyConstant<Value> {
```

Comment:
What's the advantage of using this property wrapper instead of a simple `lazy var` ? Do you have any specific use case here ?

Reply:
@notbenoit 🤦‍♂ this should actually be removed, it's a piece of an experiment I forgot to remove. However, to answer your question, the issue is that `lazy var` implies that because it's a `var`, nothing prevents you to change it value after. For example, that could happen:
```swift
final class Foo {
	private let foobar = 0
	lazy var bar = self.foobar

	func update() {
		self.bar = 1
	}
}

let foo = Foo()
foo.update()
print(foo.bar) // Prints 1
```
I think that most of the use case of using `lazy var` is to avoid writing (using the example above):
```swift
final class Foo {
	private let foobar = 0
	let bar: Int

	init() {
		self.bar = self.foobar
	}

	func update() {
		self.bar = 1 // Compile error
	}
}
```
`lazy` makes it less code, but also more error-prone, and my thinking is that we should as much as possible have everything enforce at compile time. 
Hopefully Swift one day to allow use `let bar = self.foobar` rather than having to type it out.